java/code/src/com/redhat/rhn/frontend/action/systems/SPMigrationAction.java

```diff
@@ -454,4 +459,78 @@ private SUSEProductSet createProductSet(Long baseProduct, Long[] addonProducts)
         }
         return channelIDs;
     }
+
+    /**
+     * Gets list of available migration targets for given ProductSet
+     *
+     * @param request
+     * @param ctx
+     * @param server
+     * @param installedProducts
+     * @return
+     */
+    private List<SUSEProductSet> getMigrationTargets(HttpServletRequest request, RequestContext ctx, Server server,
+                                                     Optional<SUSEProductSet> installedProducts) {
+        List<SUSEProductSet> allMigrationTargets = DistUpgradeManager.
+                getTargetProductSets(installedProducts,
+                        server.getServerArch().getCompatibleChannelArch(),
+                        ctx.getCurrentUser());
+
+        Optional<Set<String>> missingSuccessorExtensions = Optional.of(new HashSet<String>());
+        List<SUSEProductSet> migrationTargets = DistUpgradeManager.removeIncompatibleTargets(installedProducts,
+                allMigrationTargets, missingSuccessorExtensions);
+        request.setAttribute(MISSING_SUCCESSOR_EXTENSIONS,
+                missingSuccessorExtensions.orElse(new HashSet<String>()));
+        return migrationTargets;
+    }
+
+    /**
+     * Populate the request attributes from a given action
+     *
+     * @param request the HttpServletRequest
+     * @param ctx the RequestContext
+     * @return String containing the ActionForward target
+     */
+    private String populateRequestFromAction(HttpServletRequest request, RequestContext ctx) {
+        Long aid = ctx.getParamAsLong("aid");
+
+        DistUpgradeAction action = (DistUpgradeAction) ActionFactory.lookupById(aid);
+        DistUpgradeActionDetails details = action.getDetails();
+        List<Channel> channels = details.getChannelTasks().stream()
+                .filter(channel -> channel.getTask() == DistUpgradeChannelTask.SUBSCRIBE)
+                .map(DistUpgradeChannelTask::getChannel)
+                .collect(Collectors.toList());
+
+        Set<Channel> baseChannelSet = channels.stream()
+                .filter(Channel::isBaseChannel)
+                .collect(Collectors.toSet());
+
+        if (baseChannelSet.size() != 1) {
+            return TARGET;
+        }
+
+        Channel baseChannel = baseChannelSet.iterator().next();
+        List<Long> channelIds = channels.stream().map(Channel::getId).collect(Collectors.toList());
+        List<EssentialChannelDto> childChannels = getChannelDTOs(ctx, baseChannel, channelIds);
+
+        SUSEProduct baseProduct = SUSEProductFactory.lookupByChannelName(baseChannel.getName()).get(0).getRootProduct();
+
+        Server server = ctx.lookupAndBindServer();
+        Optional<SUSEProductSet> installedProducts = server.getInstalledProductSet();
+        Optional<SUSEProductSet> targetProductSet = getMigrationTargets(request, ctx, server, installedProducts)
+                .stream().filter(productSet -> productSet.getBaseProduct().equals(baseProduct)).findFirst();
```

Comment:
Can it be that there are multiple valid migration target with the same base product? Is selecting the first one the right thing to do?